src/psyclone/psyir/transformations/loop_tiling_2d_trans.py

```diff
@@ -0,0 +1,150 @@
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: S. Siso, STFC Daresbury Lab
+# -----------------------------------------------------------------------------
+
+'''This module provides the LoopTiling2DTrans, which transforms a 2D Loop
+construct into a tiled implementation of the construct.'''
+
+from psyclone.psyir.nodes import Loop
+from psyclone.psyir.transformations.chunk_loop_trans import ChunkLoopTrans
+from psyclone.psyir.transformations.loop_swap_trans import LoopSwapTrans
+from psyclone.psyir.transformations.loop_trans import LoopTrans
+from psyclone.psyir.transformations.transformation_error import \
+        TransformationError
+
+
+class LoopTiling2DTrans(LoopTrans):
+    '''
+    Apply a 2D loop tiling transformation to a loop. For example:
+
+    >>> from psyclone.psyir.frontend.fortran import FortranReader
+    >>> from psyclone.psyir.nodes import Loop
+    >>> from psyclone.psyir.transformations import LoopTiling2DTrans
+    >>> psyir = FortranReader().psyir_from_source("""
+    ... subroutine sub()
+    ...     integer :: ji, tmp(100)
+    ...     do i=1, 100
+    ...       do j=1, 100
+    ...         tmp(i, j) = 2 * tmp(i, j)
+    ...       enddo
+    ...     enddo
+    ... end subroutine sub""")
+    >>> loop = psyir.walk(Loop)[0]
+    >>> LoopTiling2DTrans().apply(loop)
+
+    will generate:
+
+    .. code-block:: fortran
+
+        subroutine sub()
+            integer :: ji
+            integer, dimension(100) :: tmp
+            integer :: ji_el_inner
+            integer :: ji_out_var
+            do i_out_var = 1, 100, 32
+              i_el_inner = MIN(i_out_var + (32 - 1), 100)
+              do j_out_var = 1, 100, 32
+                do i = i_out_var, i_el_inner, 1
+                  j_el_inner = MIN(j_out_var + (32 - 1), 100)
+                  do j = j_out_var, j_el_inner, 1
+                    tmp(i, j) = 2 * tmp(i, j)
+                  enddo
+                enddo
+              enddo
+            enddo
+        end subroutine sub
+
+    '''
+    def __str__(self):
+        return "Tile the loop construct using 2D blocks"
+
+    def validate(self, node, options=None):
+        '''
+        Validates that the given Loop node can have a LoopTiling2DTrans
+        applied.
+
+        :param node: the loop to validate.
+        :type node: :py:class:`psyclone.psyir.nodes.Loop`
+        :param options: a dict with options for transformation.
+        :type options: dict of str:values or None
+        :param int options["tilesize"]: The size to tile size for this \
```

Comment:
This assumes that we always want square tiles. I can't see any immediate reason to worry about this now though.

Reply:
Yes, I clarified this in the docstring.